src/Compilers/Core/Portable/SourceGeneration/SourceGeneratorContext.cs

```diff
@@ -76,6 +76,9 @@ internal SourceGeneratorContext(Compilation compilation, ImmutableArray<Addition
         /// The severity of the diagnostic may cause the compilation to fail, depending on the <see cref="Compilation"/> settings.
         /// </remarks>
         public void ReportDiagnostic(Diagnostic diagnostic) => _diagnostics.Add(diagnostic);
+
+        internal (ImmutableArray<GeneratedSourceText> sources, ImmutableArray<Diagnostic> diagnostics) ToImmutableAndFree()
+            => (_additionalSources.ToImmutableAndFree(), _diagnostics.ToReadOnlyAndFree());
```

Comment:
This is a user accessible type. What happens when they incorrectly stash this type aaway and use it after this method is called? Seems like it will just null ref. Do we want to include any checks here to give a better `Exception` when it's used in an invalid way? 

Not sure if there is a good parallel to analyzers here.

Reply:
Hmm, I mean potentially bad things can start happening when you store state in a generator anyway. We don't make any guarantees about lifetime/state management etc, so doing so is already off the correct path.

We could update the additional sources to throw an 'InvalidOperationException` or something explicit if you try and add sources after it's been disposed, but I'm not sure if its actually worth it?

---
In reply to: [460217283](https://github.com/dotnet/roslyn/pull/46296#discussion_r460217283) [](ancestors = 460217283)